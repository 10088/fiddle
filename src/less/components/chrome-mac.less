.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
