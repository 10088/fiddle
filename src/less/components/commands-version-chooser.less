button.version-chooser.bp3-button {
  border-radius: 3px;
}
