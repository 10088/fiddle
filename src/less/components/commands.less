@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
