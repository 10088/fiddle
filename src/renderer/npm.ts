import { EditorValues } from '../interfaces';
import { exec } from '../utils/exec';
import { fancyImport } from '../utils/import';

export interface NpmOperationOptions {
  dir: string;
}

/**
 * Finds npm modules in editor values, returning an array of modules.
 *
 * @param {EditorValues} values
 * @returns {Array<string>}
 */
export async function findModulesInEditors(values: EditorValues): Promise<Array<string>> {
  const files = [values.main, values.renderer];
  const modules: Array<string> = [];

  for (const file of files) {
    const fileModules = await findModules(file);
    modules.push(...fileModules);
  }

  return modules;
}

/**
 * Uses a simple regex to find `require()` statements in a string.
 * Tries to exclude electron and Node built-ins as well as file-path
 * references.
 *
 * @param {string} input
 * @returns {Array<string>}
 */
export async function findModules(input: string): Promise<Array<string>> {
  const matchRequire = /require\(['"]{1}([\w\d\/\-\_]*)['"]{1}\)/;
  const matched = input.match(matchRequire);
  const result: Array<string> = [];
  const builtinModules = (await fancyImport('builtin-modules') as any).default;

  if (matched && matched.length > 0) {
    const candidates = matched.slice(1);
    candidates.forEach((candidate) => {
      if (candidate === 'electron') return;
      if (builtinModules.indexOf(candidate) > -1) return;
      if (candidate.startsWith('.')) return;
      result.push(candidate);
    });
  }

  console.log(`findModules: Result`, result);

  return result;
}

/**
 * Installs given modules to a given folder.
 *
 * @param {NpmOperationOptions} { dir }
 * @param {...Array<string>} names
 * @returns {Promise<string>}
 */
export function installModules({ dir }: NpmOperationOptions, ...names: Array<string>): Promise<string> {
  return exec(dir, [ 'npm i' ].concat([ '-S' ], names).join(' '));
}

/**
 * Execute an "npm run" command
 *
 * @param {NpmOperationOptions} { dir }
 * @param {string} command
 * @returns {Promise<string>}
 */
export function npmRun({ dir }: NpmOperationOptions, command: string): Promise<string> {
  return exec(dir, `npm run ${command}`);
}
