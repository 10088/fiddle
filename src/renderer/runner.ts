import * as os from 'os';
import * as path from 'path';
import isRunning from 'is-running';
import { ChildProcess } from 'child_process';
import { Stream } from 'stream';

import {
  EditorValues,
  FileTransform,
  RunResult,
  RunnableVersion,
} from '../interfaces';
import { IpcEvents } from '../ipc-events';
import { PackageJsonOptions } from '../utils/get-package';
import { maybePlural } from '../utils/plural-maybe';
import { ipcRendererManager } from './ipc';
import {
  findModulesInEditors,
  getIsPackageManagerInstalled,
  installModules,
  packageRun,
  PMOperationOptions,
} from './npm';
import { AppState } from './state';
import {
  FiddleFactory,
  Installer,
  Paths as RunnerPaths,
  Runner as FiddleRunner,
  SpawnSyncOptions,
  SpawnOptions,
} from 'electron-fiddle-runner';
import { USER_DATA_PATH } from './constants';

export enum ForgeCommands {
  PACKAGE = 'package',
  MAKE = 'make',
}

export class Runner {
  public child: ChildProcess | null = null;
  private runner: FiddleRunner;
  private readonly fiddleFactory: FiddleFactory;

  constructor(
    private readonly appState: AppState,
    private readonly installer: Installer,
  ) {
    const paths: Partial<RunnerPaths> = {
      versionsCache: path.join(USER_DATA_PATH, 'versions.json'),
      fiddles: path.join(os.tmpdir(), 'electron-fiddles'),
    };
    this.fiddleFactory = new FiddleFactory(paths.fiddles);
    FiddleRunner.create({
      fiddleFactory: this.fiddleFactory,
      installer: this.installer,
      paths,
    }).then((runner) => (this.runner = runner));

    this.run = this.run.bind(this);
    this.stop = this.stop.bind(this);

    ipcRendererManager.removeAllListeners(IpcEvents.FIDDLE_RUN);
    ipcRendererManager.removeAllListeners(IpcEvents.FIDDLE_PACKAGE);
    ipcRendererManager.removeAllListeners(IpcEvents.FIDDLE_MAKE);

    ipcRendererManager.on(IpcEvents.FIDDLE_RUN, this.run);
    ipcRendererManager.on(IpcEvents.FIDDLE_PACKAGE, () => {
      this.performForgeOperation(ForgeCommands.PACKAGE);
    });
    ipcRendererManager.on(IpcEvents.FIDDLE_MAKE, () => {
      this.performForgeOperation(ForgeCommands.MAKE);
    });
  }

  /**
   * Bisect the current fiddle across the specified versions.
   *
   * @param {Array<RunnableVersion>} versions - versions to bisect
   * @returns {Promise<RunResult>}
   * @memberof Runner
   */
  public autobisect(versions: Array<RunnableVersion>): Promise<RunResult> {
    const { appState } = this;
    appState.isAutoBisecting = true;
    const done = () => (appState.isAutoBisecting = false);
    return this.autobisectImpl(versions).finally(done);
  }

  /**
   * Bisect the current fiddle across the specified versions.
   *
   * @param {Array<RunnableVersion>} versions - versions to bisect
   * @returns {Promise<RunResult>}
   * @memberof Runner
   */
  public async autobisectImpl(
    versions: Array<RunnableVersion>,
  ): Promise<RunResult> {
    const prefix = `Runner: autobisect`;
    const { appState } = this;

    // precondition: can't bisect unless we have >= 2 versions
    if (versions.length < 2) {
      appState.pushOutput(`${prefix} needs at least two Electron versions`);
      return RunResult.INVALID;
    }

    // show the console
    if (appState.isClearingConsoleOnRun) appState.clearConsole();
    appState.isConsoleShowing = true;

    // FIXME: install packages

    // autobisect
    const { pushOutput } = this.appState;
    const out = new Stream.PassThrough();
    out.on('readable', () => pushOutput(out.read()));
    const opts: SpawnSyncOptions = { out, showConfig: false };
    appState.isRunning = true;
    await this.runner.bisect(
      versions[0].version,
      versions[versions.length - 1].version,
      Object.entries(await window.ElectronFiddle.app.getEditorValues()),
      opts,
    );
    appState.isRunning = false;

    return RunResult.SUCCESS;
  }

  /**
   * Actually run the fiddle.
   *
   * @returns {Promise<RunResult>}
   */
  public async run(): Promise<RunResult> {
    const { appState } = this;
    const { state, version } = appState.currentElectronVersion;

    if (state !== 'installed') {
      console.warn(`Runner: Binary ${version} not ready`);
      const message = `Please wait for ${version} to finish installing before running the fiddle.`;
      appState.pushOutput(message, { isNotPre: true });
      return RunResult.INVALID;
    }

    if (appState.isClearingConsoleOnRun) appState.clearConsole();
    appState.isConsoleShowing = true;

    return this.execute();
  }

  /**
   * Stop a currently running Electron fiddle.
   *
   * @returns {boolean} true if runner is now idle
   * @memberof Runner
   */
  public stop(): void {
    this.appState.isRunning = !!this.child && !this.child.kill();

    // If the child process is still alive 1 second after we've
    // attempted to kill it by normal means, kill it forcefully.
    setTimeout(() => {
      const pid = this.child?.pid;
      if (pid && isRunning(pid)) {
        this.child?.kill('SIGKILL');
      }
    }, 1000);
  }

  /**
   * Uses electron-forge to either package or make the current fiddle
   *
   * @param {ForgeCommands} operation
   * @returns {Promise<boolean>}
   * @memberof Runner
   */
  public async performForgeOperation(
    operation: ForgeCommands,
  ): Promise<boolean> {
    const options = { includeDependencies: true, includeElectron: true };
    const { dotfilesTransform } = await import('./transforms/dotfiles');
    const { forgeTransform } = await import('./transforms/forge');
    const { pushError, pushOutput } = this.appState;

    const strings =
      operation === ForgeCommands.MAKE
        ? ['Creating installers for', 'Binary']
        : ['Packaging', 'Installers'];

    this.appState.isConsoleShowing = true;
    pushOutput(`ðŸ“¦ ${strings[0]} current Fiddle...`);

    const packageManager = this.appState.packageManager;
    const pmInstalled = await getIsPackageManagerInstalled(packageManager);
    if (!pmInstalled) {
      let message = `Error: Could not find ${packageManager}. Fiddle requires Node.js and npm or yarn `;
      message += `to compile packages. Please visit https://nodejs.org to install `;
      message += `Node.js and npm, or https://classic.yarnpkg.com/lang/en/ `;
      message += `to install Yarn`;

      this.appState.pushOutput(message, { isNotPre: true });
      return false;
    }

    // Save files to temp
    const dir = await this.saveToTemp(
      options,
      dotfilesTransform,
      forgeTransform,
    );
    if (!dir) return false;

    // Files are now saved to temp, let's install Forge and dependencies
    if (!(await this.packageInstall({ dir, packageManager }))) return false;

    // Cool, let's run "package"
    try {
      console.log(`Now creating ${strings[1].toLowerCase()}...`);
      pushOutput(await packageRun({ dir, packageManager }, operation));
      pushOutput(`âœ… ${strings[1]} successfully created.`, { isNotPre: true });
    } catch (error) {
      pushError(`Creating ${strings[1].toLowerCase()} failed.`, error);
      return false;
    }

    const { shell } = await import('electron');
    shell.showItemInFolder(path.join(dir, 'out'));
    return true;
  }

  /**
   * Analyzes the editor's JavaScript contents for modules
   * and installs them.
   *
   * @param {EditorValues} values
   * @param {string} dir
   * @returns {Promise<void>}
   */
  public async installModulesForEditor(
    values: EditorValues,
    pmOptions: PMOperationOptions,
  ): Promise<void> {
    const modules = await findModulesInEditors(values);
    const { pushOutput } = this.appState;

    if (modules && modules.length > 0) {
      this.appState.isInstallingModules = true;
      const packageManager = pmOptions.packageManager;
      const pmInstalled = await getIsPackageManagerInstalled(packageManager);
      if (!pmInstalled) {
        let message = `The ${maybePlural(`module`, modules)} ${modules.join(
          ', ',
        )} need to be installed, `;
        message += `but we could not find ${packageManager}. Fiddle requires Node.js and npm `;
        message += `to support the installation of modules not included in `;
        message += `Electron. Please visit https://nodejs.org to install Node.js `;
        message += `and npm, or https://classic.yarnpkg.com/lang/en/ to install Yarn`;

        pushOutput(message, { isNotPre: true });
        this.appState.isInstallingModules = false;
        return;
      }

      pushOutput(
        `Installing node modules using ${
          pmOptions.packageManager
        }: ${modules.join(', ')}...`,
        { isNotPre: true },
      );

      const result = await installModules(pmOptions, ...modules);
      pushOutput(result);

      this.appState.isInstallingModules = false;
    }
  }

  private buildChildEnvVars(): { [x: string]: string | undefined } {
    const { isEnablingElectronLogging, environmentVariables } = this.appState;

    const env = { ...process.env };

    if (isEnablingElectronLogging) {
      env.ELECTRON_ENABLE_LOGGING = 'true';
      env.ELECTRON_DEBUG_NOTIFICATIONS = 'true';
      env.ELECTRON_ENABLE_STACK_DUMPING = 'true';
    } else {
      delete env.ELECTRON_ENABLE_LOGGING;
      delete env.ELECTRON_DEBUG_NOTIFICATIONS;
      delete env.ELECTRON_ENABLE_STACK_DUMPING;
    }

    for (const v of environmentVariables) {
      const [key, value] = v.split('=');
      env[key] = value;
    }

    return env;
  }

  /**
   * Execute Electron.
   *
   * @param {string} dir
   * @param {string} version
   * @returns {Promise<void>}
   * @memberof Runner
   */
  public async execute(): Promise<RunResult> {
    // spawn the process
    const {
      currentElectronVersion,
      flushOutput,
      // isKeepingUserDataDirs,
      pushOutput,
    } = this.appState;
    const { version, localPath } = currentElectronVersion;

    // create the fiddle snapshot
    const { getEditorValues } = window.ElectronFiddle.app;
    const values = await getEditorValues();
    const entries = Object.entries(values) as Iterable<[string, string]>;
    const fiddle = await this.fiddleFactory.create(entries);

    const env = this.buildChildEnvVars();
    const opts: SpawnOptions = { env, showConfig: false };

    const child = await this.runner.spawn(localPath || version, fiddle!, opts);

    this.child = child;
    this.appState.isRunning = true;
    pushOutput(`Electron v${version} started.`);

    return new Promise((resolve) => {
      const onData = (str: string) => pushOutput(str, { bypassBuffer: false });
      child.stdout!.on('data', onData);
      child.stderr!.on('data', onData);
      child.on('close', async (code, signal) => {
        flushOutput();
        this.appState.isRunning = false;
        this.child = null;
        // FIXME if (!isKeepingUserDataDirs) await fiddle!.remove();
        if (typeof code !== 'number') {
          pushOutput(`Electron exited with signal ${signal}.`);
          resolve(RunResult.INVALID);
        } else {
          pushOutput(`Electron exited with code ${code}.`);
          resolve(!code ? RunResult.SUCCESS : RunResult.FAILURE);
        }
      });
    });
  }

  /**
   * Save files to temp, logging to the Fiddle terminal while doing so
   *
   * @param {PackageJsonOptions} options
   * @param {...Array<FileTransform>} transforms
   * @returns {(Promise<string | null>)}
   * @memberof Runner
   */
  public async saveToTemp(
    options: PackageJsonOptions,
    ...transforms: Array<FileTransform>
  ): Promise<string | null> {
    const { fileManager } = window.ElectronFiddle.app;
    const { pushOutput, pushError } = this.appState;

    try {
      pushOutput(`Saving files to temp directory...`);
      const dir = await fileManager.saveToTemp(options, ...transforms);
      pushOutput(`Saved files to ${dir}`);
      return dir;
    } catch (error) {
      pushError('Failed to save files.', error.message);
    }

    return null;
  }

  /**
   * Installs modules in a given directory (we're basically
   * just running "{packageManager} install")
   *
   * @param {PMOperationOptions} options
   * @returns
   * @memberof Runner
   */
  public async packageInstall(options: PMOperationOptions): Promise<boolean> {
    try {
      this.appState.pushOutput(`Now running "npm install..."`);
      this.appState.pushOutput(await installModules(options));
      return true;
    } catch (error) {
      this.appState.pushError('Failed to run "npm install".', error);
    }

    return false;
  }
}
