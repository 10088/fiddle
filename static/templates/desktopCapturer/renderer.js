const { desktopCapturer } = require('electron')

// The following example shows how to capture video from
// the screen. It also grabs each window, so you could
// just grab video from a single window.
//
desktopCapturer.getSources({ types: ['window', 'screen'] }, (error, sources) => {
  if (error) throw error
  for (let i = 0; i < sources.length; ++i) {
    console.log(sources[i])

    if (sources[i].name.startsWith('Screen')) {
      navigator.mediaDevices.getUserMedia({
        audio: false,
        video: {
          mandatory: {
            chromeMediaSource: 'desktop',
            chromeMediaSourceId: sources[i].id,
            minWidth: 1280,
            maxWidth: 1280,
            minHeight: 720,
            maxHeight: 720
          }
        }
      })
      .then((stream) => handleStream(stream))
      .catch((e) => handleError(e))
      return
    }
  }
})

function handleStream (stream) {
  const video = document.querySelector('video')
  video.srcObject = stream
  video.onloadedmetadata = (e) => video.play()
}

function handleError (e) {
  console.log(e)
}