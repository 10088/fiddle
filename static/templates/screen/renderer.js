const { screen } = require('electron')

const screens = screen.getAllDisplays()
const div = document.querySelector('div')

div.innerText = JSON.stringify(screens);
