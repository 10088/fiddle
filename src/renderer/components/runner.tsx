import { ChildProcess, spawn } from 'child_process';
import { observer } from 'mobx-react';
import * as path from 'path';
import * as React from 'react';

import { EditorValues, FileTransform } from '../../interfaces';
import { IpcEvents } from '../../ipc-events';
import { PackageJsonOptions } from '../../utils/get-package';
import { normalizeVersion } from '../../utils/normalize-version';
import { maybePlural } from '../../utils/plural-maybe';
import { ipcRendererManager } from '../ipc';
import { findModulesInEditors, getIsNpmInstalled, installModules, npmRun } from '../npm';
import { AppState } from '../state';

export interface RunnerState {
  isRunning: boolean;
}

export interface RunnerProps {
  appState: AppState;
}

export enum ForgeCommands {
  PACKAGE = 'package',
  MAKE = 'make'
}

/**
 * The runner component is responsible for actually launching the fiddle
 * with Electron. It also renders the button that does so.
 *
 * @class Runner
 * @extends {React.Component<RunnerProps, RunnerState>}
 */
@observer
export class Runner extends React.Component<RunnerProps, RunnerState> {
  public child: ChildProcess | null = null;

  constructor(props: RunnerProps) {
    super(props);

    this.run = this.run.bind(this);
    this.performForgeOperation = this.performForgeOperation.bind(this);
    this.stop = this.stop.bind(this);
    this.state = { isRunning: false };

    this.props.appState.pushOutput('Console ready ðŸ”¬');
  }

  public componentDidMount() {
    ipcRendererManager.on(IpcEvents.FIDDLE_RUN, this.run);
    ipcRendererManager.on(IpcEvents.FIDDLE_PACKAGE, () => {
      this.performForgeOperation(ForgeCommands.PACKAGE);
    });
    ipcRendererManager.on(IpcEvents.FIDDLE_MAKE, () => {
      this.performForgeOperation(ForgeCommands.MAKE);
    });
  }

  public render() {
    const { versions, version } = this.props.appState;
    const { isRunning } = this.state;

    if (!versions || !version || !versions[normalizeVersion(version)]) {
      return null;
    }

    const state = versions[normalizeVersion(version)].state;

    let text = 'Run';
    let action: () => any = this.run;

    if (state === 'downloading') {
      text = 'Downloading';
    }

    if (isRunning) {
      text = 'Stop';
      action = this.stop;
    }

    return (
      <button
        className='button button-run'
        onClick={() => action()}
      >
          {text}
      </button>
    );
  }

  /**
   * Stop a currently running Electron fiddle.
   */
  public async stop() {
    if (this.child) {
      this.child.kill();
      this.setState({
        isRunning: false
      });
    }
  }

  /**
   * Analyzes the editor's JavaScript contents for modules
   * and installs them.
   *
   * @param {EditorValues} values
   * @param {string} dir
   * @returns {Promise<void>}
   */
  public async installModulesForEditor(values: EditorValues, dir: string): Promise<void> {
    const modules = await findModulesInEditors(values);
    const { appState } = this.props;

    if (modules && modules.length > 0) {
      if (!(await getIsNpmInstalled())) {
        let message = `The ${maybePlural(`module`, modules)} ${modules.join(', ')} need to be installed, `;
        message += `but we could not find npm. Fiddle requires Node.js and npm `;
        message += `to support the installation of modules not included in `;
        message += `Electron. Please visit https://nodejs.org to install Node.js `;
        message += `and npm.`;

        appState.pushOutput(message, { isNotPre: true });
        return;
      }

      appState.pushOutput(`Installing npm modules: ${modules.join(', ')}...`, { isNotPre: true });
      appState.pushOutput(await installModules({ dir }, ...modules));
    }
  }

  /**
   * Execute Electron.
   *
   * @param {string} dir
   * @param {string} version
   * @returns {Promise<void>}
   * @memberof Runner
   */
  public async execute(dir: string): Promise<void> {
    const { appState } = this.props;
    const { version, pushOutput } = appState;

    const binaryPath = appState.binaryManager.getElectronBinaryPath(version);
    console.log(`Runner: Binary ${binaryPath} ready, launching`);

    this.child = spawn(binaryPath, [ dir, '--inspect' ]);
    this.setState({ isRunning: true });
    pushOutput(`Electron v${version} started.`);

    this.child.stdout.on('data', (data) => pushOutput(data, { bypassBuffer: false }));
    this.child.stderr.on('data', (data) => pushOutput(data, { bypassBuffer: false }));
    this.child.on('close', (code) => {
      const withCode = typeof code === 'number'
        ? ` with code ${code.toString()}.`
        : `.`;

      pushOutput(`Electron exited${withCode}`);
      this.setState({ isRunning: false });
      this.child = null;
      window.ElectronFiddle.app.fileManager.cleanup(dir);
    });
  }

  /**
   * Save files to temp, logging to the Fiddle terminal while doing so
   *
   * @param {PackageJsonOptions} options
   * @param {...Array<FileTransform>} transforms
   * @returns {(Promise<string | null>)}
   * @memberof Runner
   */
  public async saveToTemp(
    options: PackageJsonOptions, ...transforms: Array<FileTransform>
  ): Promise<string | null> {
    const { fileManager } = window.ElectronFiddle.app;
    const { pushOutput, pushError } = this.props.appState;

    try {
      pushOutput(`Saving files to temp directory...`);
      const dir = await fileManager.saveToTemp(options, ...transforms);
      pushOutput(`Saved files to ${dir}`);
      return dir;
    } catch (error) {
      pushError('Failed to save files.', error);
    }

    return null;
  }

  /**
   * Installs modules in a given directory (we're basically
   * just running "npm install")
   *
   * @param {string} dir
   * @returns
   * @memberof Runner
   */
  public async npmInstall(dir: string): Promise<boolean> {
    try {
      this.props.appState.pushOutput(`Now running "npm install..."`);
      this.props.appState.pushOutput(await installModules({ dir }));
      return true;
    } catch (error) {
      this.props.appState.pushError('Failed to run "npm install".', error);
    }

    return false;
  }

  /**
   * Uses electron-forge to either package or make the current fiddle
   *
   * @param {ForgeCommands} operation
   * @returns {Promise<boolean>}
   * @memberof Runner
   */
  public async performForgeOperation(operation: ForgeCommands): Promise<boolean> {
    const options = { includeDependencies: true, includeElectron: true };
    const { dotfilesTransform } = await import('../transforms/dotfiles');
    const { forgeTransform } = await import('../transforms/forge');
    const { appState } = this.props;
    const { pushError, pushOutput } = appState;

    const strings = operation === ForgeCommands.MAKE
      ? [ 'Creating installers for', 'Binary' ]
      : [ 'Packaging', 'Installers' ];

    appState.isConsoleShowing = true;
    pushOutput(`ðŸ“¦ ${strings[0]} current Fiddle...`);

    if (!(await getIsNpmInstalled())) {
      let message = `Error: Could not find npm. Fiddle requires Node.js and npm `;
      message += `to compile packages. Please visit https://nodejs.org to install `;
      message += `Node.js and npm.`;

      appState.pushOutput(message, { isNotPre: true });
      return false;
    }

    // Save files to temp
    const dir = await this.saveToTemp(options, dotfilesTransform, forgeTransform);
    if (!dir) return false;

    // Files are now saved to temp, let's install Forge and dependencies
    if (!(await this.npmInstall(dir))) return false;

    // Cool, let's run "package"
    try {
      console.log(`Now creating ${strings[1].toLowerCase()}...`);
      pushOutput(await npmRun({ dir }, operation));
      pushOutput(`âœ… ${strings[1]} successfully created.`, { isNotPre: true });
    } catch (error) {
      pushError(`Creating ${strings[1].toLowerCase()} failed.`, error);
      return false;
    }

    const { shell } = await import('electron');
    shell.showItemInFolder(path.join(dir, 'out'));
    return true;
  }

  /**
   * Actually run the fiddle.
   *
   * @returns {Promise<boolean>}
   */
  public async run(): Promise<boolean> {
    const { appState } = this.props;
    const { fileManager, getValues } = window.ElectronFiddle.app;
    const options = { includeDependencies: false, includeElectron: false };
    const { binaryManager, version } = appState;

    appState.isConsoleShowing = true;

    const isDownloaded = await binaryManager.getIsDownloaded(version);
    const values = await getValues(options);
    const dir = await this.saveToTemp(options);

    if (!dir) return false;

    try {
      await this.installModulesForEditor(values, dir);
    } catch (error) {
      console.error('Runner: Could not install modules', error);
      fileManager.cleanup(dir);
      return false;
    }

    if (!isDownloaded) {
      console.warn(`Runner: Binary ${version} not ready`);

      let message = `Could not start fiddle: `;
      message += `Electron ${version} not downloaded yet. `;
      message += `Please wait for it to finish downloading `;
      message += `before running the fiddle.`;

      appState.pushOutput(message, { isNotPre: true });
      fileManager.cleanup(dir);
      return false;
    }

    this.execute(dir);

    return true;
  }
}
