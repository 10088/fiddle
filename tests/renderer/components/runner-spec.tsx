import * as React from 'react';
import { shallow } from 'enzyme';
import { spawn } from 'child_process';

import { Runner, ForgeCommands } from '../../../src/renderer/components/runner';
import { mockVersions } from '../../mocks/electron-versions';
import { ElectronFiddleMock } from '../../mocks/electron-fiddle';
import { findModulesInEditors, installModules, npmRun } from '../../../src/renderer/npm';
import { MockChildProcess } from '../../mocks/child-process';

jest.mock('../../../src/renderer/npm');
jest.mock('electron', () => require('../../mocks/electron'));
jest.mock('fs-extra');
jest.mock('child_process');

describe('Runner component', () => {
  let mockChild: MockChildProcess;

  beforeEach(() => {
    mockChild = new MockChildProcess();

    this.store = {
      version: '2.0.2',
      versions: mockVersions,
      downloadVersion: jest.fn(),
      removeVersion: jest.fn(),
      pushOutput: jest.fn(),
      pushError: jest.fn(),
      binaryManager: {
        getIsDownloaded: jest.fn(() => true),
        getElectronBinaryPath: jest.fn((version: string) => `/fake/path/${version}/electron`)
      },
    };

    (window as any).ElectronFiddle = new ElectronFiddleMock();
  });

  it('renders', () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('runs', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    (findModulesInEditors as any).mockReturnValueOnce([ 'fake-module' ]);
    (spawn as any).mockReturnValueOnce(mockChild);

    expect(await instance.run()).toBe(true);
    expect(this.store.binaryManager.getIsDownloaded).toHaveBeenCalled();
    expect(window.ElectronFiddle.app.fileManager.saveToTemp).toHaveBeenCalled();
    expect(installModules).toHaveBeenCalled();
    expect(wrapper.state('isRunning')).toBe(true);
  });

  it('emits output', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    (findModulesInEditors as any).mockReturnValueOnce([ 'fake-module' ]);
    (spawn as any).mockReturnValueOnce(mockChild);

    // Output
    expect(await instance.run()).toBe(true);
    mockChild.stdout.emit('data', 'hi');
    mockChild.stderr.emit('data', 'hi');
    expect(this.store.pushOutput).toHaveBeenCalledTimes(8);

    // Stop
    mockChild.emit('close', 0);
    expect(wrapper.state('isRunning')).toBe(false);
  });

  it('stops on close', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    (findModulesInEditors as any).mockReturnValueOnce([ 'fake-module' ]);
    (spawn as any).mockReturnValueOnce(mockChild);

    // Stop
    expect(await instance.run()).toBe(true);
    expect(wrapper.state('isRunning')).toBe(true);
    instance.stop();
    expect(wrapper.state('isRunning')).toBe(false);
  });

  it('stops on stop()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    (findModulesInEditors as any).mockReturnValueOnce([ 'fake-module' ]);
    (spawn as any).mockReturnValueOnce(mockChild);

    // Stop
    expect(await instance.run()).toBe(true);
    mockChild.emit('close', 0);
    expect(wrapper.state('isRunning')).toBe(false);
  });

  it('does not run version not yet downloaded', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    this.store.binaryManager.getIsDownloaded.mockReturnValueOnce(false);

    expect(await instance.run()).toBe(false);
  });

  it('does not run if writing files fails', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    (window.ElectronFiddle.app.fileManager.saveToTemp as jest.Mock)
      .mockImplementationOnce(() => {
        throw new Error('bwap bwap');
      });

    expect(await instance.run()).toBe(false);
  });

  it('installs modules on installModules()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    expect(await instance.npmInstall('')).toBe(true);
    expect(installModules).toHaveBeenCalled();
  });

  it('handles an error in installModules()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;
    (installModules as any).mockImplementationOnce(() => {
      throw new Error('bwap bwap');
    });

    expect(await instance.npmInstall('')).toBe(false);
    expect(installModules).toHaveBeenCalled();
  });

  it('performs a package operation in performForgeOperation()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    expect(await instance.performForgeOperation(ForgeCommands.PACKAGE)).toBe(true);
  });

  it('performs a make operation in performForgeOperation()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;

    expect(await instance.performForgeOperation(ForgeCommands.MAKE)).toBe(true);
  });

  it('handles an error in saveToTemp() in performForgeOperation()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;
    (instance as any).saveToTemp = jest.fn();

    expect(await instance.performForgeOperation(ForgeCommands.MAKE)).toBe(false);
  });

  it('handles an error in npmInstall() in performForgeOperation()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;
    (installModules as any).mockImplementationOnce(() => {
      throw new Error('bwap bwap');
    });

    expect(await instance.performForgeOperation(ForgeCommands.MAKE)).toBe(false);
  });

  it('handles an error in npmRun() in performForgeOperation()', async () => {
    const wrapper = shallow(<Runner appState={this.store} />);
    const instance: Runner = wrapper.instance() as any;
    (npmRun as any).mockImplementationOnce(() => {
      throw new Error('bwap bwap');
    });

    expect(await instance.performForgeOperation(ForgeCommands.MAKE)).toBe(false);
  });
});
