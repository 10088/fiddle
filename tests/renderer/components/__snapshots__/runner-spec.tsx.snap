// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Runner component renders 1`] = `
<button
  className="button button-run"
  onClick={[Function]}
>
  Run
</button>
`;
